### DISCUSSION TITLE
Unable to locate attachment issue

### DISCUSSION THREAD

**Discussion by roshansridhar** [ReagentX/imessage-exporter#402] [Q&A] [answered]
I have many old messages on iMessage where the attachments haven’t been downloaded. When I export and check the HTML threads, some images are missing, and I often see "Unable to locate attachment:" in their place. How can I retrieve all the old attachments without having to take hours to go thru every conversation and manually select every photo.?

<img width="440" alt="Screenshot 2024-12-08 at 6 29 39 PM" src="https://github.com/user-attachments/assets/f24221ab-8ec2-4676-a854-d5bb857617bf">

**Comment by ReagentX** [comment, marked as answer]
In the Messages app, if you click the info button for a conversation and scroll down, there is a button to force it to download all of the attachments for that conversation:

<img width="321" alt="image" src="https://github.com/user-attachments/assets/8fd8e7a2-1d8c-46a5-9d8e-8645317e2458">

**Reply by mpk99** [reply]
Didnt want to hijack the convo, but this doesn't quite work for me.  It says there are 12 items in iCloud, but there are many more in a convo i'm trying to archive.  I'm struggling as i have to scroll back to like 2 yrs ago, slowly, as it only loads like 1/2 a day at a time, and when i find the attacment i want, if i click "download now" in the imessage chat, it downloads, but that's the only way i can get all the old attachments to show up.  I'm not sure if i'm doing something wrong, but there must be another way.

btw this program rocks, the issue i'm describing is an imessage issue, not one for your fantastic program.

**Reply by ReagentX** [reply]
I agree that it doesn't always work; I have used [SendKeys](https://github.com/socsieng/sendkeys) to continually send the `pgup` keystroke to scroll conversations back semi-manually.

**Reply by mpk99** [reply]
wow what a quick response.  I'll check it out, seems straight forward, just have to figure out how to make it target a specific conversation

**Reply by ReagentX** [reply]
If you double-click a conversation in the list view it will open it in a new window.

**Reply by mpk99** [reply]
Yeah i got it somewhat sort of working.  Now my next issue, after i get about a few good 100 pgup clicks, it reverts back to the bottom of the chat 🤦 fml

**Reply by HolisticDeveloper** [reply]
I take it that the `Download` link something in the MacOS Messages app -- I don't see anything similar on the iPhone.

**Reply by ReagentX** [reply]
Here is a screenshot of how it looks in my
iPhone, at the bottom of the conversation information view that opens when you tap the contact's name at the top of the screen: 

![IMG_9204](https://github.com/user-attachments/assets/5e238626-744b-44e7-916d-a8d09f42319a)

If all the images have already been downloaded, the button may not appear.

**Reply by HolisticDeveloper** [reply]
> Here is a screenshot of how it looks in my iPhone

I think I just have this option turned off. (Under Settings > Apple Account > iCloud > Saved to iCloud, Messages is "Off.")

In my first run with imessage-exporter, I had many occurrences of missing attachments. Searching led me here, so I thought it was an issue with offloading to iCloud. I guess I had forgotten I didn't have the setting enabled on the iPhone. (It's the only Apple device I own so I probably didn't see a good reason to enable it.)

After reviewing the old message thread on my phone and taking a new backup, imessage-exporter was able to find them. Not sure why that would be. I'll keep experimenting to see what I can figure out.

**Comment by roshansridhar** [comment]
Thank you so much! Is there a way to avoid doing this for each conversation individually? I noticed there might be an option in your tool to passing it as a parameter external db or attachment folders. Would obtaining this using third-party apps be a viable solution?

**Reply by ReagentX** [reply]
I don't know of any third party apps that can force a synchronization with iCloud.

**Comment by Meatwad650** [comment]
Forcing download fixed the issue for me, but oddly I was able to find the attachments before that and they were the right attachments.
